### PR TITLE
Implement fmpz_mod_mat.

### PR DESCRIPTION
**Pull request by wbhart**
This implements fmpz_mod_mat, modulo problems that already exist in nmod_mat (I fixed some of them, but not all).

However, it requires the latest commit in flint/trunk and there is no test code or documentation.

I have tested all the functions by hand, so they should work, but I have run out of time to write the test code and will be away next week. If you are interested in taking over writing the test code @thofma then go for it. Also the docs should probably mention the new module and its type and that it is provided by flint.

Some issues I discovered with nmod_mat:

* the setindex, set_entry mess
* the transpose option can now be done with Julia transposes
* I doubt the code for setting a matrix to an fmpz_mat works
* I doubt deepcopy works
* negative powers are not accepted
* no test for square matrix in powering
* there is no (trivial) constructor S(M) where S is a matrix space and M is a matrix in that space
* addeq! is missing
* is tr really the right name for trace?
* isone is missing
* there is no need to reduce the output of functions like trace, one can construct an nmod directly
* functions like zero_matrix, identity_matrix, sub, view, probably need updating in line with recent changes we made in AbstractAlgebra
* other things I forgot.....

### CODE REVIEW

**Comment by thofma** [comment]
Howell and strong echelon form are implemented in https://github.com/wbhart/flint2/blob/trunk/fmpz_mat/howell_form_mod.c and https://github.com/wbhart/flint2/blob/trunk/fmpz_mat/strong_echelon_form_mod.c

**Comment by thofma** [comment]
There is also an rref function: https://github.com/wbhart/flint2/blob/trunk/fmpz_mat/rref_mod.c

**Comment by thofma** [comment]
I guess I we should wait with new binaries till these things are sorted out on the flint side?

**Comment by wbhart** [comment]
I didn't even realise we could use those. I assumed from the names that they were modular algorithms for computing the fmpz_mat functions.

I guess these should be moved or wrapped somehow.

**Comment by thofma** [comment]
https://github.com/wbhart/flint2/pull/592

**Comment by wbhart** [comment]
I've added tests and changed the docs. Of course it wont pass until binaries are updated for Flint.

**Comment by thofma** [comment]
Just to have it in writing, https://github.com/Nemocas/Nemo.jl/issues/676 and https://github.com/Nemocas/Nemo.jl/issues/676 are blocking this.

**Comment by thofma** [comment]
Also the tests are failing.

**Comment by wbhart** [comment]
Of course the tests will fail until the binaries are updated.

**Comment by wbhart** [comment]
Also #676 and #676 are the same item.

**Comment by thofma** [comment]
#677

**Comment by wbhart** [comment]
Thanks. I'm not sure if I misread that.

**Comment by thofma** [comment]
No, you were right, I edited it after I saw your comment.

File: src/flint/fmpz_mod_mat.jl [hunk 1]
```diff
@@ -0,0 +1,863 @@
...
+end
+
+base_ring(a::FmpzModMatSpace) = a.base_ring
+
+base_ring(a::T) where T <: Zmod_fmpz_mat = a.base_ring
+
+zero(a::FmpzModMatSpace) = a()
+
+function one(a::FmpzModMatSpace)
+  (nrows(a) != ncols(a)) && error("Matrices must be quadratic")
```
Comment (rfourquet) [inline src/flint/fmpz_mod_mat.jl]: I was not aware that "quadratic matrix" means "square matrix", "square" might be preferable for non-native speakers.

Comment (wbhart) [reply src/flint/fmpz_mod_mat.jl]: It doesn't mean that. This is something that German people write. If I put it there, it is a copy-paste error. I'll fix it.

File: src/flint/fmpz_mod_mat.jl [hunk 2]
```diff
@@ -0,0 +1,863 @@
...
+          (Ref{T}, Ref{T}, Int), z, x, y)
+  return z
+end
+=#
+
+function ^(x::T, y::fmpz) where T <: Zmod_fmpz_mat
+  (y > fmpz(typemax(Int))) &&
+          error("Exponent must be smaller then ", fmpz(typemax(Int)))
+  (y < fmpz(typemin(Int))) &&
+          error("Exponent must be bigger then ", fmpz(typemin(Int)))
```
Comment (rfourquet) [inline src/flint/fmpz_mod_mat.jl]: Is "then" an official valid substitute for "than" (just asking, I always wondered as I see this form so often). 
I suggest merging these two checks into one, e.g. 
```julia
typemin(Int) <= y <= typemax(Int) ||
    throw(DomainError(y, "exponent must be in $(typemin(Int):typemax(Int))"))
```

**Comment by rfourquet** [comment]
Side comment: I see that you used 2 spaces indentation here, which has precedent in Nemo but otherwise it's mostly 3-spaces. It might be useful to consolidate into one style, would you mind if I do that eventually? (I can tell my editor to use a specific indentation for a specific project, but I can't tell it to use mixed indentation depending on the context in a given file).

**Comment by wbhart** [comment]
I think that's just because I copy and pasted from nmod_mat.jl. I don't mind if you fix the indents to be consistent, though it is very low priority.

File: src/flint/fmpz_mod_mat.jl (hunk 2 above)
Comment (wbhart) [reply src/flint/fmpz_mod_mat.jl]: Fixed.

File: src/flint/fmpz_mod_mat.jl (hunk 1 above)
Comment (wbhart) [reply src/flint/fmpz_mod_mat.jl]: Fixed.

**Comment by wbhart** [comment]
The current generic code is no longer correct. So I don't think 676 and 677 should hold this up any more, as the problem is already there and this PR doesn't make it any worse.

**Comment by thofma** [comment]
Agreed.

**Pull request merged**